### PR TITLE
Add joint MMD

### LINKED ISSUE #1109: Add JMMD
### What's the new feature?

Add JMMD from https://arxiv.org/abs/2504.10139

### What value does this add?

Allows for discrepancies between joint distributions to be measured.

### Is there an alternative you've considered?

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dominicjbroadbent**
### PR Type

- Feature
- Tests

### Description

Closes #1109 

### How Has This Been Tested?

Adding tests according to format of `test_metrics.py`.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have ensured my code is easy to understand, including docstrings and comments where necessary.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have updated CHANGELOG.md, if appropriate.

### CODE REVIEW

File: coreax/metrics.py [hunk 1]
```diff
@@ -427,3 +427,117 @@ def compute(
...
+        comparison_data: SupervisedData,
+        **kwargs,
+    ) -> Array:
+        r"""
+        Compute the (weighted) joint maximum mean discrepancy.
+
+        .. math::
+            \text{JMMD}^2(\mathcal{D}_1,\mathcal{D}_2) = \mathbb{E}(k(\mathcal{D}_1,
+            \mathcal{D}_1)) + \mathbb{E}(k(\mathcal{D}_2,\mathcal{D}_2))
+            - 2\mathbb{E}(k(\mathcal{D}_1,\mathcal{D}_2))
```
Comment (tm91236) [inline coreax/metrics.py]: Should these use `r` instead of `k` to match the class docstring?

File: tests/unit/test_metrics.py [hunk 2]
```diff
@@ -534,3 +534,218 @@ def test_negative_regularisation_parameter(self):
...
+            l(\mathcal{D}_1,\mathcal{D}_2)) w_2
+
+            = (\frac{37}{16} + \frac{12}{16} * e^{-1})
+            + (\frac{13}{36} - \frac{1}{3} * e^{-4})
+            - 2 * (\frac{1}{8} + \frac{6}{8} * e^{-1} - \frac{1}{12} * e^{-4} -
+            \frac{6}{12} * e^{-9})
+        """
+        reference_data = jnp.array([[-1], [0], [1]])
+        reference_supervision = jnp.array([[-1], [0], [1]])
+        reference_weights = jnp.array([1 / 3, 1 / 6, 1 / 2])
```
Comment (tm91236) [inline tests/unit/test_metrics.py]: To improve the test quality, please can you instead use weights that do not sum to one? This way we test if any explicit normalisation is required.

**Review by tm91236 — CHANGES_REQUESTED** [review]
Thanks for the contribution! Generally looks good, I've left a few minor comments.

File: coreax/metrics.py [hunk 3]
```diff
@@ -427,3 +427,104 @@ def compute(
...
+        .. math::
+            \text{JMMD}^2(\mathcal{D}_1,\mathcal{D}_2) = \mathbb{E}(r(\mathcal{D}_1,
+            \mathcal{D}_1)) + \mathbb{E}(r(\mathcal{D}_2,\mathcal{D}_2))
+            - 2\mathbb{E}(r(\mathcal{D}_1,\mathcal{D}_2))
+
+        :param reference_data: Supervised dataset :math:`\mathcal{D}_1 =
+            \{(x_i, y_i)\}_{i=1}^n` with :math:`x \in \mathbb{R}^d` and
+            :math:`y \in \mathbb{R}^p`
+        :param comparison_data: Supervised dataset
+            :math:`\mathcal{D}_2 = \{(x^\prime_i, y^\prime_i)\}_{i=1}^n` with
```
Comment (tm91236) [inline coreax/metrics.py]: Should this be `\mathcal{D}_2 = \{(x^\prime_i, y^\prime_i)\}_{i=1}^m` instead?

**Review by tm91236 — CHANGES_REQUESTED** [review]
Thank you for the changes. Please can you update the CHANGELOG to include the addition of JMMD and the now merged AMCMD (this was missed in #1112), and address the minor comments.

**Review by tm91236 — CHANGES_REQUESTED** [review]
Just a few nit picks remain, otherwise LGTM!

File: coreax/metrics.py [hunk 4]
```diff
@@ -427,3 +427,104 @@ def compute(
...
+    :math:`\mathcal{D}^{(2)} = \{(x^\prime_i, y^\prime_i)\}_{i=1}^m`
+    with :math:`x^\prime\in\mathbb{R}^d` and :math:`y^\prime\in\mathbb{R}^p`,
+    the joint maximum mean discrepancy is given by:
+
+    .. math::
+        \text{JMMD}^2(\mathcal{D}_1,\mathcal{D}_2) = \mathbb{E}(r(\mathcal{D}_1,
+        \mathcal{D}_1)) + \mathbb{E}(r(\mathcal{D}_2,\mathcal{D}_2))
+        - 2\mathbb{E}(r(\mathcal{D}_1,\mathcal{D}_2))
+
+    where :math:`r` is a tensor-product kernel, defined here as the product of a feature
```
[the remark this answers is not in the document]
Comment (dominicjbroadbent) [reply coreax/metrics.py]: Agreed, changed!

File: coreax/metrics.py [hunk 5]
```diff
@@ -427,3 +427,104 @@ def compute(
...
+    kernel and a response kernel, and the expectation is with respect to
+    the normalised data weights.
+
+    .. note::
+        Assuming that the tensor-product kernel is characteristic
+        (:cite:`muandet2017rkhs`), it can be shown that
+        :math:`\text{JMMD}^2(\mathcal{D}_1,\mathcal{D}_2) = 0` if and only if
+        :math:`\mathbb{P}^{(1)}_(X, Y) = \mathbb{P}^{(2)}_(X, Y)`, i.e. the joint
```
[the remark this answers is not in the document]
Comment (dominicjbroadbent) [reply coreax/metrics.py]: Fixed!

File: coreax/metrics.py [hunk 6]
```diff
@@ -427,3 +427,104 @@ def compute(
...
+
+    feature_kernel: ScalarValuedKernel
+    response_kernel: ScalarValuedKernel
+
+    def compute(
+        self,
+        reference_data: SupervisedData,
+        comparison_data: SupervisedData,
+        **kwargs,
+    ) -> Array:
```
[the remark this answers is not in the document]
Comment (dominicjbroadbent) [reply coreax/metrics.py]: Yes, fixed!

File: tests/unit/test_metrics.py [hunk 7]
```diff
@@ -534,3 +534,234 @@ def test_negative_regularisation_parameter(self):
...
+        output = metric.compute(reference_dataset, comparison_dataset)
+        assert output == pytest.approx(expected_output)
+
+    def test_jmmd_analytically_known_weighted(self) -> None:
+        r"""
+        Test JMMD computation against an analytically derived weighted solution.
+
+        Weighted jmmd is calculated if and only if comparison_weights are provided. When
+        `comparison_weights` = :data:`None`, the JMMD class computes the standard,
+        non-weighted MMD.
```
[the remark this answers is not in the document]
Comment (dominicjbroadbent) [reply tests/unit/test_metrics.py]: This is not true, removed it

**Comment by dominicjbroadbent** [comment]
Apologies for the delay on this, was on holiday

**Review by tm91236 — APPROVED** [review]
